colcon_ros_domain_id_coordinator/shell/ros_domain_id.py

```diff
@@ -70,19 +80,44 @@ def create_hook_prepend_value(  # noqa: D102
     async def generate_command_environment(  # noqa: D102
         self, task_name, build_base, dependencies,
     ):
-        try:
-            domain_id = self.free_ids.get_nowait()
-        except QueueEmpty:
-            logger.warn(f"No free ROS_DOMAIN_ID to assign for '{task_name}'")
+        # Allocate unique IDs based on the build_base
+        build_base = str(build_base)
+
+        if build_base in self.allocated_ids:
+            domain_id = self.allocated_ids[build_base]
+        else:
+            try:
+                domain_id = self.free_ids.get_nowait()
+            except QueueEmpty:
+                domain_id = None
+
+        if domain_id is None:
+            logger.warn(f"No free ROS_DOMAIN_ID to assign for '{build_base}'")
             os.environ.pop('ROS_DOMAIN_ID', None)
-            domain_id = None
         else:
             os.environ['ROS_DOMAIN_ID'] = domain_id
             logger.debug(
-                f"Allocated ROS_DOMAIN_ID={domain_id} for '{task_name}'")
+                f"Allocated ROS_DOMAIN_ID={domain_id} for '{build_base}'")
 
-            # Place the ID at the end of the FIFO to be reused if needed
-            self.free_ids.put_nowait(domain_id)
+        self.allocated_ids[build_base] = domain_id
```

Comment:
It is, I'll add a comment.
There are some tasks which will call this shell function multiple times and we want to be consistent. If we failed the first time but succeeded a subsequent time, it could yield some really confusing behavior.

Reply:
Done in a4cee4f.